File: 백준/dp/11053 가장 긴 증가하는 부분수열.py
N = int(input())
arr = list(map(int, input().split(" ")))

def solution():
    ans = [1] * N

    for i in range(N):
        for j in range(i):
            if arr[i] > arr[j]:
                ans[i] = max(ans[i], ans[j]+1)

    print(max(ans))

solution()

File: 백준/dp/11053가장긴증가하는부분수열.js
const solution = (M, numList) => {
  const dp = [];
  for (let i = 0; i < M; i++) {
    dp.push(1);
  }
  for (let i = 0; i < M; i++) {
    for (let j = 0; j < i; j++) {
      if (numList[i] > numList[j]) {
        dp[i] = Math.max(dp[i], dp[j] + 1);
      }
    }
  }
  console.log(Math.max(...dp));
};

const fs = require('fs');

const stdin = (process.platform === 'linux'
  ? fs.readFileSync('/dev/stdin').toString()
  : `6
10 20 10 30 20 50
  `
).split('\n');

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const M = parseInt(input(), 10);
const numList = input().split(' ').map(Number);

solution(M, numList);
